lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,343 @@
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* global getComputedStyle, getElementsInDocument, Node, getNodePath, getNodeSelector, window */
+
+const Gatherer = require('../gatherer');
+const pageFunctions = require('../../../lib/page-functions.js');
+const {rectContainsString, rectContains} = require('../../../lib/rect-helpers');
+
+const TARGET_SELECTORS = [
+  'button',
+  'a',
+  'input',
+  'textarea',
+  'select',
+  'option',
+  '[role=button]',
+  '[role=checkbox]',
+  '[role=link]',
+  '[role=menuitem]',
+  '[role=menuitemcheckbox]',
+  '[role=menuitemradio]',
+  '[role=option]',
+  '[role=scrollbar]',
+  '[role=slider]',
+  '[role=spinbutton]',
+];
+const tapTargetsSelector = TARGET_SELECTORS.join(',');
+
+/**
+ * @param {Element} element
+ * @returns {boolean}
+ */
+/* istanbul ignore next */
+function elementIsVisible(element) {
+  const {overflowX, overflowY, display, visibility} = getComputedStyle(element);
+
+  if (
+    display === 'none' ||
+    (visibility === 'collapse' && ['TR', 'TBODY', 'COL', 'COLGROUP'].includes(element.tagName))
+  ) {
+    // Element not displayed
+    return false;
+  }
+
+  // only for block and inline-block, since clientWidth/Height are always 0 for inline elements
+  if (display === 'block' || display === 'inline-block') {
+    // if height/width is 0 and no overflow in that direction then
+    // there's no content that the user can see and tap on
+    if ((element.clientWidth === 0 && overflowX === 'hidden') ||
+        (element.clientHeight === 0 && overflowY === 'hidden')) {
+      return false;
+    }
+  }
+
+  const parent = element.parentElement;
+  if (parent && parent.tagName !== 'BODY') {
+    // if a parent is invisible then the current element is also invisible
+    return elementIsVisible(parent);
+  }
+
+  return true;
+}
+
+/**
+ * @param {LH.Artifacts.Rect[]} clientRects
+ */
+/* istanbul ignore next */
+function allClientRectsEmpty(clientRects) {
+  return clientRects.every(cr => cr.width === 0 && cr.height === 0);
+}
+
+/**
+ * @param {Element} element
+ */
+/* istanbul ignore next */
+function getVisibleClientRects(element) {
+  if (!elementIsVisible(element)) {
+    return [];
+  }
+
+  let clientRects = getClientRects(element);
+
+  if (allClientRectsEmpty(clientRects)) {
+    const {
+      overflowX,
+      overflowY,
+    } = getComputedStyle(element);
+    if ((overflowX === 'hidden' && overflowY === 'hidden') || element.children.length === 0) {
```

Comment:
if all the client rects are size 0, does it matter whether or not the children would be visible since their client rects are also size 0 (since they're included in the `allClientRectsEmpty(clientRects)` check)?

Reply:
No it shouldn't matter, and from trying it out on a few sites it doesn't matter. Will remove the if.